tcms/rpc/api/testcase.py

```diff
@@ -279,8 +284,37 @@ def filter(query=None):  # pylint: disable=redefined-builtin
     if query is None:
         query = {}
 
-    return list(
-        TestCase.objects.filter(**query)
+    for key, val in query.items():
+        if not key.startswith(
+            ("setup_duration", "testing_duration", "expected_duration")
+        ):
+            continue
+
+        try:
+            duration = parse_duration(val)
+        except TypeError:
+            # val isn't a string or byte-like object
+            # item is probably something like 'setup_duration__isnull=True'
+            continue
+
+        if duration is None:
+            continue
+
+        query[key] = duration
+
+    qs = (
+        TestCase.objects.annotate(
+            expected_duration=Case(
+                When(
+                    Q(setup_duration__isnull=True) & Q(testing_duration__isnull=True),
+                    then=timedelta(0),
+                ),
+                When(Q(setup_duration__isnull=True), then="testing_duration"),
+                When(Q(testing_duration__isnull=True), then="setup_duration"),
+                default=F("setup_duration") + F("testing_duration"),
+            )
+        )
+        .filter(**query)
```

Comment:
It looks like there may be some confusion around this item. Being able to filter by the duration field is the last checklist item in #1923. That should not need special casing for setup_duration and testing_duration because these are model fields. The only special case is expected_duration. 

OTOH the item "API method TestCase.filter returns the value of all 3 fields - tests are updated to match" refers only to making sure that the API method TestCase.filter includes the new fields in its results. That should be much easier and require a lot less changes.

Reply:
Oh oh. I was over thinking it.

I thought the current checklist item also meant that filtering with respect to the duration fields should be possible  on the API side.

I will create a new PR to implement just the presence of the duration fields in the result.

I'll keep this one alive (closed) so one can refer to it if we ever need to add filtering with respect to the duration fields on the API side.

Reply:
It just occured to me that I can shorten the entire `Case` expression with `Coalesce`:

```Python
Coalesce("setup_duration", timedelta(0)) + Coalesce("testing_duration", timedelta(0))
```